mongo.go

```diff
@@ -0,0 +1,149 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"sync"
+	"time"
+
+	_ "github.com/lib/pq"
+	log "github.com/sirupsen/logrus"
+)
+
+type mongoConfig struct {
+	host       string
+	replicaSet string
+	port       int
+	user       string
+	authSource string
+	password   string
+	database   string
+	caCert     string
+	tls        bool
+	clientCert string
+	clientKey  string
+}
+
+func (mongo mongoConfig) dump(sb s3Backend, keyPath, database string) error {
+	today := time.Now().Format("20060102150405")
+	mongo.database = database
+	mongoURI := buildMongoURI(mongo)
+	log.Debugln(mongoURI)
+	cmd := exec.Command("mongodump", mongoURI, "--archive")
+
+	var out bytes.Buffer
+	cmd.Stdout = &out
+
+	var errMsg bytes.Buffer
+	cmd.Stderr = &errMsg
+
+	err := cmd.Run()
+	if err != nil {
+		log.Errorf(errMsg.String())
+		return err
+	}
+
+	wg := sync.WaitGroup{}
+	wr, err := sb.NewFileWriter(today+"-"+database+".archive", &wg)
+	if err != nil {
+		log.Errorf("Could not open backup file for writing: %v", err)
+		return err
+	}
+
+	key := getKey(keyPath)
+	e, err := newEncryptor(key, wr)
+	if err != nil {
+		log.Errorf("Could not initialize encryptor: (%v)", err)
+		return err
+	}
+
+	c, err := newCompressor(key, e)
+	if err != nil {
+		log.Errorf("Could not initialize compressor: (%v)", err)
+		return err
+	}
+
+	_, err = c.Write(out.Bytes())
+	if err != nil {
+		log.Errorf("Could not encrypt/write: %s", err)
+		return err
+	}
+
+	c.Close()
+	wr.Close()
+	wg.Wait()
+
+	return nil
+}
+
+func (mongo mongoConfig) restore(sb s3Backend, keyPath, archive string) error {
+
+	fr, err := sb.NewFileReader(archive)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	defer fr.Close()
+
+	key := getKey(keyPath)
+	r, err := newDecryptor(key, fr)
+	if err != nil {
+		log.Error("Could not initialise decryptor", err)
+		return err
+	}
+	d, err := newDecompressor(key, r)
+	if err != nil {
+		log.Error("Could not initialise decompressor", err)
+		return err
+
+	}
+	data, err := ioutil.ReadAll(d)
```

Comment:
`ReadFrom` into `in` instead? (Avoids having an extra copy needing to be present in memory, or one can go for a pipe to save even more :).